src/Microsoft.PowerShell.Security/security/CertificateProvider.cs

```diff
@@ -3340,17 +3341,53 @@ private DnsNameRepresentation GetDnsNameRepresentation(string dnsName)
-
-            // extract DNS name from subject distinguish name
-            // if it exists and does not contain a comma
-            // a comma, indicates it is not a DNS name
-            if (cert.Subject.StartsWith(distinguishedNamePrefix, StringComparison.OrdinalIgnoreCase) &&
-                !cert.Subject.Contains(','))
+            // Extract DNS name from Subject distinguished name
+            foreach (X500RelativeDistinguishedName rdn in cert.SubjectName.EnumerateRelativeDistinguishedNames())
             {
-                string parsedSubjectDistinguishedDnsName = cert.Subject.Substring(distinguishedNamePrefix.Length);
-                DnsNameRepresentation dnsName = GetDnsNameRepresentation(parsedSubjectDistinguishedDnsName);
-                _dnsList.Add(dnsName);
+                // Extract Common Name directly if RDN has single attribute which we detect when HasMultipleElements == false
+                // This is the common case which .NET supports directly
+                if (!rdn.HasMultipleElements)
+                {
+                    string oid = rdn.GetSingleElementType().Value;
+                    string rdnValue = rdn.GetSingleElementValue();
+
+                    if (!string.IsNullOrEmpty(rdnValue) &&
+                        CommonNameOid.Equals(oid, StringComparison.Ordinal))
+                    {
+                        DnsNameRepresentation dnsName = GetDnsNameRepresentation(rdnValue);
+                        _dnsList.Add(dnsName);
+                        break;
+                    }
+                }
+
+                // Handle edge case with multi-value RDN when HasMultipleElements == true
+                // We go back to AsnReader and parse the Common Name
+                // .NET does not support this directly but provides the raw data in the RDN so we can parse it ourselves
+                else
+                {
+                    AsnReader asnReader = new(rdn.RawData, AsnEncodingRules.DER);
+                    AsnReader setReader = asnReader.ReadSetOf(skipSortOrderValidation: true);
+
+                    while (setReader.HasData)
+                    {
+                        AsnReader elementReader = setReader.ReadSequence();
+                        string oid = elementReader.ReadObjectIdentifier();
+
+                        if (CommonNameOid.Equals(oid, StringComparison.Ordinal))
+                        {
+                            Asn1Tag tag = elementReader.PeekTag();
+
+                            if (tag.TagClass == TagClass.Universal)
+                            {
+                                string decodedValue = elementReader.ReadCharacterString((UniversalTagNumber)tag.TagValue);
+                                DnsNameRepresentation dnsName = GetDnsNameRepresentation(decodedValue);
+                                _dnsList.Add(dnsName);
+                            }
+
+                            break;
+                        }
```

Comment:
I probably need to get all multi-value CNs here actually. This breaks on the first one found.

Reply:
Actually not sure about this one. Might need to investigate a bit more.